src/librustc_trans/back/msvc/mod.rs

```diff
@@ -42,110 +44,133 @@ mod platform {
     use std::path::{Path, PathBuf};
     use std::process::Command;
     use session::Session;
-    use super::registry::{LOCAL_MACHINE};
+    use super::arch::{host_arch, Arch};
+    use super::registry::LOCAL_MACHINE;
 
-    // Cross toolchains depend on dlls from the host toolchain
-    // We can't just add it to the Command's PATH in `link_exe_cmd` because it
-    // is later overridden so we publicly expose it here instead
-    pub fn host_dll_path() -> Option<PathBuf> {
-        get_vc_dir().and_then(|(_, vcdir)| {
-            host_dll_subdir().map(|sub| {
-                vcdir.join("bin").join(sub)
+    // First we need to figure out whether the environment is already correctly
+    // configured by vcvars. We do this by looking at the environment variable
+    // `VCINSTALLDIR` which is always set by vcvars, and unlikely to be set
+    // otherwise. If it is defined, then we find `link.exe` in `PATH and trust
+    // that everything else is configured correctly.
+    //
+    // If `VCINSTALLDIR` wasn't defined (or we couldn't find the linker where
+    // it claimed it should be), then we resort to finding everything
+    // ourselves. First we find where the latest version of MSVC is installed
+    // and what version it is. Then based on the version we find the
+    // appropriate SDKs.
+    //
+    // For versions are are unable to detect the user has to execute the
+    // appropriate vcvars bat file themselves to configure the environment.
+    //
+    // If despite our best efforts we are still unable to find MSVC then we
+    // just blindly call `link.exe` and hope for the best.
+    pub fn link_exe_cmd(sess: &Session) -> (Command, Option<PathBuf>) {
+        let arch = &sess.target.target.arch;
+        env::var_os("VCINSTALLDIR").and_then(|_| {
+            debug!("Detected that vcvars was already run.");
+            env::var_os("PATH").and_then(|path| {
+                // Mingw has its own link which is not the linkwe want so we
+                // look for `cl.exe` too as a precaution.
+                env::split_paths(&path).find(|path| {
+                    path.join("cl.exe").is_file()
+                        && path.join("link.exe").is_file()
+                }).map(|path| {
+                    (Command::new(path.join("link.exe")), None)
+                })
             })
+        }).or_else(|| {
+            find_msvc_14(arch).or_else(|| {
```

Comment:
I think I've asked you this before, but this chain of if/else makes me think that we're going to have to touch up this logic for all releases of Visual Studio, is that right? I think in the past you said they were done because of the ucrt, but does that mean that none of the other directories will ever change either?